File: shared/scoring.ts
import {
  HookItem,
  HookEvaluation,
  RankedHook,
  HookScoreRubric,
} from './types';

export const SCORING_WEIGHTS = {
  scrollStop: 0.30,
  relatability: 0.25,
  curiosityGap: 0.20,
  clarity: 0.15,
  brandFit: 0.10,
} as const;

/**
 * Calculates weighted score from the 5 rubric criteria.
 * Formula: Scroll-stop .30, Relatability .25, Curiosity .20, Clarity .15, Brand fit .10
 * Returns score rounded to 2 decimal places.
 */
export function calculateWeightedScore(scores: HookScoreRubric): number {
  const raw =
    scores.scrollStop * SCORING_WEIGHTS.scrollStop +
    scores.relatability * SCORING_WEIGHTS.relatability +
    scores.curiosityGap * SCORING_WEIGHTS.curiosityGap +
    scores.clarity * SCORING_WEIGHTS.clarity +
    scores.brandFit * SCORING_WEIGHTS.brandFit;

  return Math.round(raw * 100) / 100;
}

/**
 * Ranks all hooks based on their weighted rubric score.
 * Enforces compliance guardrails:
 * If a hook has compliance flags and a safe rewrite, the displayText is set to the safe rewrite.
 * Top 3 are elevated with isTop3 = true.
 */
export function rankHooks(
  hooks: HookItem[],
  evaluations: HookEvaluation[]
): RankedHook[] {
  const evalMap = new Map<string, HookEvaluation>();
  evaluations.forEach((e) => evalMap.set(e.hookId, e));

  const ranked: RankedHook[] = hooks.map((h) => {
    const evaluation = evalMap.get(h.id) || {
      hookId: h.id,
      scores: {
        scrollStop: 5,
        relatability: 5,
        clarity: 5,
        curiosityGap: 5,
        brandFit: 5,
      },
      critique: 'Standard evaluation pending.',
      rewrite: null,
      compliance: {
        isCompliant: true,
        flags: [],
        safeRewrite: null,
      },
    };

    const weightedScore = calculateWeightedScore(evaluation.scores);
    const hasFlags =
      !evaluation.compliance.isCompliant ||
      evaluation.compliance.flags.length > 0;
    const safeRewrite = evaluation.compliance.safeRewrite;

    // If compliance flag exists, use safe rewrite for display if provided
    const displayText = hasFlags && safeRewrite ? safeRewrite : h.text;

    return {
      hookId: h.id,
      rank: 1, // temporary placeholder
      rawText: h.text,
      displayText,
      weightedScore,
      scores: evaluation.scores,
      critique: evaluation.critique,
      rewrite: evaluation.rewrite,
      compliance: evaluation.compliance,
      isTop3: false,
    };
  });

  // Sort descending by weightedScore
  ranked.sort((a, b) => b.weightedScore - a.weightedScore);

  // Assign ranks & top 3 flag
  return ranked.map((hook, index) => ({
    ...hook,
    rank: index + 1,
    isTop3: index < 3,
  }));
}

File: agent/app/scoring.py
from typing import Dict, List, Any

SCORING_WEIGHTS: Dict[str, float] = {
    "scroll_stop": 0.30,
    "relatability": 0.25,
    "curiosity_gap": 0.20,
    "clarity": 0.15,
    "brand_fit": 0.10,
}


def calculate_composite_score(scores: Dict[str, float], is_compliant: bool = True) -> float:
    """Calculates weighted composite score (0-10) matching shared/scoring.ts."""
    if not is_compliant:
        return 0.0

    scroll = scores.get("scroll_stop", scores.get("scrollStop", 0.0))
    relatability = scores.get("relatability", 0.0)
    curiosity = scores.get("curiosity_gap", scores.get("curiosityGap", scores.get("curiosity", 0.0)))
    clarity = scores.get("clarity", 0.0)
    brand = scores.get("brand_fit", scores.get("brandFit", 0.0))

    composite = (
        scroll * SCORING_WEIGHTS["scroll_stop"]
        + relatability * SCORING_WEIGHTS["relatability"]
        + curiosity * SCORING_WEIGHTS["curiosity_gap"]
        + clarity * SCORING_WEIGHTS["clarity"]
        + brand * SCORING_WEIGHTS["brand_fit"]
    )
    return round(composite, 2)


def rank_evaluated_hooks(
    hooks: List[Dict[str, Any]],
    evaluations: List[Dict[str, Any]],
    compliance_items: List[Dict[str, Any]],
) -> List[Dict[str, Any]]:
    """Merges hooks, evaluations, and compliance to produce ranked hooks."""
    eval_map = {e["hookId"]: e for e in evaluations if "hookId" in e}
    comp_map = {c["hookId"]: c for c in compliance_items if "hookId" in c}

    ranked: List[Dict[str, Any]] = []

    for hook in hooks:
        h_id = hook["id"]
        ev = eval_map.get(h_id, {})
        scores = ev.get("scores", {
            "scrollStop": 7.0,
            "relatability": 7.0,
            "curiosityGap": 7.0,
            "clarity": 7.0,
            "brandFit": 7.0,
        })
        comp = comp_map.get(h_id, {"isCompliant": True, "flags": [], "safeRewrite": None})
        is_compliant = comp.get("isCompliant", comp.get("passes", True))

        composite = calculate_composite_score(scores, is_compliant=is_compliant)

        ranked.append({
            "hookId": h_id,
            "id": h_id,
            "rank": 0,
            "rawText": hook.get("text", ""),
            "displayText": comp.get("safeRewrite") or hook.get("text", ""),
            "text": hook.get("text", ""),
            "angleIndex": hook.get("angleIndex", 0),
            "weightedScore": composite,
            "compositeScore": composite,
            "scores": {
                "scrollStop": float(scores.get("scrollStop", scores.get("scroll_stop", 7.0))),
                "relatability": float(scores.get("relatability", 7.0)),
                "curiosityGap": float(scores.get("curiosityGap", scores.get("curiosity_gap", 7.0))),
                "clarity": float(scores.get("clarity", 7.0)),
                "brandFit": float(scores.get("brandFit", scores.get("brand_fit", 7.0))),
            },
            "critique": ev.get("critique", ev.get("reasoning", "Strong conversational hook.")),
            "rewrite": comp.get("safeRewrite", None),
            "compliance": {
                "isCompliant": is_compliant,
                "flags": comp.get("flags", []),
                "safeRewrite": comp.get("safeRewrite", None),
            },
            "isTop3": False,
            "isTopHook": False,
        })

    # Sort descending by weightedScore, then relatability, then scrollStop
    ranked.sort(
        key=lambda x: (
            x["weightedScore"],
            x["scores"]["relatability"],
            x["scores"]["scrollStop"],
        ),
        reverse=True,
    )

    # Assign 1-indexed rank and mark top 3
    for i in range(len(ranked)):
        ranked[i]["rank"] = i + 1
        if i < 3:
            ranked[i]["isTop3"] = True
            ranked[i]["isTopHook"] = True

    return ranked
